frappe/utils/sentry.py

```diff
@@ -1,18 +1,93 @@
 import os
 import sys
+from datetime import datetime
 
+import rq
 from sentry_sdk import capture_message as sentry_capture_message
+from sentry_sdk import configure_scope
 from sentry_sdk.hub import Hub
+from sentry_sdk.integrations import Integration
 from sentry_sdk.integrations.wsgi import _make_wsgi_event_processor
 from sentry_sdk.tracing import SOURCE_FOR_STYLE
-from sentry_sdk.utils import event_from_exception
+from sentry_sdk.tracing_utils import record_sql_queries
+from sentry_sdk.utils import capture_internal_exceptions, event_from_exception
 
 import frappe
 import frappe.monitor
+from frappe.database.database import Database, EmptyQueryValues
+
+
+class FrappeIntegration(Integration):
+	identifier = "frappe"
+
+	@staticmethod
+	def setup_once():
+		real_connect = Database.connect
+		real_sql = Database.sql
+
+		def sql(self, query, values=None, *args, **kwargs):
+			hub = Hub.current
+
+			if not self._conn:
+				self.connect()
+
+			with record_sql_queries(
+				hub, self._cursor, query, values, paramstyle="pyformat", executemany=False
+			):
+				return real_sql(self, query, values or EmptyQueryValues, *args, **kwargs)
+
+		def connect(self):
+			hub = Hub.current
+			with capture_internal_exceptions():
+				hub.add_breadcrumb(message="connect", category="query")
+
+			with hub.start_span(op="db", description="connect"):
+				return real_connect(self)
+
+		Database.connect = connect
+		Database.sql = sql
+
+
+def set_sentry_context():
```

Comment:
This should run conditionally?

Also preferable it should run just before sending traceback.

This doesn't work well if frappe integration is enabled though, because that would require init-ing immediately.

Reply:
Cleaned up and tested with various scenarios
Testing with just `FRAPPE_SENTRY_DSN` set, and adding `ENABLE_SENTRY_DB_MONITORING` and/or `SENTRY_TRACING_SAMPLE_RATE`, works as expected